### ISSUE TITLE
Shell / Tab completion for atmos stacks and components

### ISSUE THREAD

**Issue by nitrocode** [cloudposse/atmos#45]
Have a question? Please checkout our [Slack Community](https://slack.cloudposse.com) or visit our [Slack Archive](https://archive.sweetops.com/). 

[![Slack Community](https://slack.cloudposse.com/badge.svg)](https://slack.cloudposse.com)

## Describe the Feature

Tab completion for atmos commands

## Expected Behavior

```
$ atmos<tab>
helm helmfile help istioctl stack terraform vendor version workflow
```

```
$ atmos terraform<tab>
apply backend clean console deploy destroy import init output plan show workspace
```

```
$ atmos terraform plan<tab>
tfstate-backend aws-sso github-runners # etc
```

```
$ atmos terraform plan github-runners --stack<tab>
gbl-root gbl-prod gbl-staging
```

## Use Case

I want the command and stacks to complete using tab instead of having to type it in each time

## Describe Ideal Solution

It's implemented

## Alternatives Considered

None

## Additional Context

* https://stackoverflow.com/questions/5570795/how-does-bash-tab-completion-work
* https://github.com/perlpunk/shell-completions

**Comment by nitrocode** [comment]
I did notice there is a new `completion` subcommand.

```
⨠ eval $(atmos completion bash)
⨠ atmos<tab>
completion  (generate the autocompletion script for the specified shell)
describe    (describe)
helmfile    (helmfile command)
help        (Help about any command)
terraform   (terraform command)
version     (version command)
```

This seems to be off

```
⨠ atmos terraform <tab>
--stack               -s                    generate  (generate)
```

That should show plan/apply/deploy/destroy

This returns a `-` when hitting tab instead of showing the list of components

```
⨠ atmos terraform plan <tab>
```

This returns another `-` when it should show the list of stacks 

```
⨠ atmos terraform plan efs --stack <tab>
```

**Comment by nitrocode** [comment]
@mcalhoun 

I reopened this ticket because I was hoping we could do something like this

```sh
# get components
atmos terraform plan <tab>
# get root stack
atmos terraform plan ecr --stack <tab>
```

**Comment by mcalhoun** [comment]
Yeah, I just realized I only implemented the base-level (atmos commands) and not stack completion. Figuring out how to do that will be a bit more tricky.

**Comment by nitrocode** [comment]
"real" (not abstract) component completion and root stack completion would be very handy here.

I also noticed that subcommands are missing.

```bash
✗ atmos terraform<tab>
generate   -- Execute 'terraform generate' commands
provision  -- This command provisions terraform components
```

I was expecting `plan`, `deploy`, `apply`, etc

---

Here's a way to get all the stacks using `yq`

```bash
✗ atmos describe stacks | yq e '. | keys'
- gbl-corp
- gbl-identity
- gbl-root
...
```

An even easier way is using it natively in atmos

```
✗ atmos list stacks
gbl-corp
gbl-identity
gbl-root
```

Here's how to get all the real components using `yq`

```bash
✗ atmos describe stacks | yq e '. | to_entries | .[].value.components.terraform | with_entries(select(.value.metadata.type != "abstract")) | keys' | grep -v '\[\]' | sort | uniq
- aws-team-roles
- aws-teams
- account
- account-map
- aws-team-roles
- github-oidc-provider
- vpc
```

**Comment by RoseSecurity** [comment]
I think this would be even easier if we had a command for `atmos list` where users can list the stacks, components, and variables associated with each stack. We could repurpose this function to provide some autocompletion for the command line parameters like you mentioned. In the meantime, would something like the following be suitable for the `terraform` options?

```go

var validArgs = []string{"init", "plan", "apply", "deploy", "destroy", "state", "import"} // and more as needed

// terraformCmd represents the base command for all terraform sub-commands
var terraformCmd = &cobra.Command{
	Use:                "terraform",
	Aliases:            []string{"tf"},
	Short:              "Execute Terraform commands",
	Long:               `This command executes Terraform commands`,
	FParseErrWhitelist: struct{ UnknownFlags bool }{UnknownFlags: true},
	ValidArgs:          validArgs,
	Run: func(cmd *cobra.Command, args []string) {
```

which produces the following:

```console
❯ atmos terraform <tab>
generate            -- Execute 'terraform generate' commands
apply               deploy              destroy             dns-delegated-move  import              init                plan                state
```

**Comment by samtholiya** [comment]
[https://github.com/cloudposse/atmos/pull/959](https://github.com/cloudposse/atmos/pull/959)<br><br>This pr will only add completions for atmos inbuilt command, stack names and components<br><br>For custom commands we will have to 

1. Fix the issue that shadows commands in case custom command includes a command that atmos already has [DEV-469](https://linear.app/cloudposse/issue/DEV-469/atmos-custom-subcommands-may-shadow-built-in-subcommands-eg-atmos)
2. We will have to probably modify the completion script that is generate so that it always defaults to dynamic suggestion instead of static suggestions that it stores for efficiency. So that we always read the config to suggest the commands

[DEV-469]: https://cloudposse.atlassian.net/browse/DEV-469?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
